# Objective:
To generate a recommendation system/model to get the best recommendation for book for a particular user.

# Data:
The dataset contains 3 tables:
* `Users` :-It contains user ID’s have been anonymized and map to integers.  demographic data is provided (location, age if available) otherwise this fields  contain null values. 
* `Books` :- Books are identified by their respective ISBN. 
* `Ratings` :-It contains book rating information. Ratings are either explicit,  expressed on a scale from 1 to 10, Or implicit expressed by 0. 




In [1]:
# Importing important libraries
import pandas as pd
import matplotlib.pyplot as plt

*Let's import all the dataset*

In [2]:
# Importing books csv file
books_df = pd.read_csv("/content/sample_data/Books.csv")
books_df.head()

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [3]:
books_df.shape

(271360, 8)

In [4]:
# Importing ratings csv file
ratings_df = pd.read_csv("/content/sample_data/Ratings.csv")
ratings_df.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [5]:
ratings_df.shape

(1149780, 3)

In [6]:
# Importing users csv file
users_df = pd.read_csv("/content/sample_data/Users.csv")
users_df.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [7]:
users_df.shape

(278858, 3)

# Explorartory data analysis

### Let's check the `users_df` table.

In [8]:
users_df.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [9]:
users_df.duplicated().sum()

0

In [10]:
users_df.Age.unique()

array([ nan,  18.,  17.,  61.,  26.,  14.,  25.,  19.,  46.,  55.,  32.,
        24.,  20.,  34.,  23.,  51.,  31.,  21.,  44.,  30.,  57.,  43.,
        37.,  41.,  54.,  42.,  50.,  39.,  53.,  47.,  36.,  28.,  35.,
        13.,  58.,  49.,  38.,  45.,  62.,  63.,  27.,  33.,  29.,  66.,
        40.,  15.,  60.,   0.,  79.,  22.,  16.,  65.,  59.,  48.,  72.,
        56.,  67.,   1.,  80.,  52.,  69.,  71.,  73.,  78.,   9.,  64.,
       103., 104.,  12.,  74.,  75., 231.,   3.,  76.,  83.,  68., 119.,
        11.,  77.,   2.,  70.,  93.,   8.,   7.,   4.,  81., 114., 230.,
       239.,  10.,   5., 148., 151.,   6., 101., 201.,  96.,  84.,  82.,
        90., 123., 244., 133.,  91., 128.,  94.,  85., 141., 110.,  97.,
       219.,  86., 124.,  92., 175., 172., 209., 212., 237.,  87., 162.,
       100., 156., 136.,  95.,  89., 106.,  99., 108., 210.,  88., 199.,
       147., 168., 132., 159., 186., 152., 102., 116., 200., 115., 226.,
       137., 207., 229., 138., 109., 105., 228., 18

In [11]:
len(users_df["User-ID"].unique())

278858

In [12]:
len(users_df["Location"].unique())

57339

* Nearly half of the `Age` observations are null/NaN.
* Also, `Age` column has many outliers such as [0, 100+ age] etc.
* `278858` unique users were found.
* `57339` unique locations were found.
* No duplicates were found

### Let's check `books_df` table

In [13]:
books_df.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [14]:
books_df.duplicated().sum()

0

In [15]:
books_df["Book-Title"].unique()

array(['Classical Mythology', 'Clara Callan', 'Decision in Normandy', ...,
       'Lily Dale : The True Story of the Town that Talks to the Dead',
       "Republic (World's Classics)",
       "A Guided Tour of Rene Descartes' Meditations on First Philosophy with Complete Translations of the Meditations by Ronald Rubin"],
      dtype=object)

In [16]:
len(books_df["Book-Title"].unique())

242135

In [17]:
books_df["Book-Author"].value_counts()

Agatha Christie         632
William Shakespeare     567
Stephen King            524
Ann M. Martin           423
Carolyn Keene           373
                       ... 
Linda Jackson             1
Artiste C. Arthur         1
Emily Pearl Kingsley      1
Jean Cassels              1
Christopher  Biffle       1
Name: Book-Author, Length: 102023, dtype: int64

In [18]:
len(books_df["Book-Author"].unique())

102024

In [19]:
books_df["Year-Of-Publication"].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

* There are also some null/Nan values present in `books_df` table. But the count of values is very low, so we can ignore them or if we don't need them later, we can remove them later as well.
* There are `50763` unique books present in the table.
* There are `25402` unique authors are present. The most books are by author Stephen King.
* The books are spread over different year-of-publication. The same column has outliers as well for eg `2030` which is yet to come. Also some NaN values and zeros as well.
* No duplicated were found.

### Let's check `ratings_df` table

In [20]:
ratings_df.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [21]:
ratings_df.duplicated().sum()

0

In [22]:
len(ratings_df["User-ID"].unique())

105283

* Only `1` books is not rated.
* `69513` unique users were found.
* No duplicates are there.

In [23]:
# Merging all the files to make it one for easy use
books_df_copy = books_df.copy()    # Making a copy of datasets so we don't work on original data.
ratings_df_copy = ratings_df.copy()
users_df_copy = users_df.copy()

books_ratings_df = pd.merge(books_df, ratings_df, how="inner", left_on="ISBN", right_on="ISBN")
final_df = pd.merge(books_ratings_df, users_df, how="inner", left_on="User-ID", right_on="User-ID")

In [24]:
final_df.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  User-ID  Book-Rating  \
0  http://images.amazon.com/images/P/0195153448.0...        2            0   
1  http://images.amazon.com/images/P/0002005018.0...        8            5   
2  http://images.amazon.com/images/P/0060973129.0...        8            0   
3  http://images.amazon.com/images/P/0374157065.0...        8            0   
4  http://images.amazon.com/images/P/0393045218.0...        8            0   

                    Location   Age  
0  stockton, california, usa  18.0  
1   timmins, ontario, canada   NaN  
2   timmins, ontario, canada   NaN  
3   timmins, ontario, canada   NaN  
4   timmins, ontario, canada   NaN

In [25]:
final_df.shape

(1031136, 12)

In [26]:
final_df.duplicated().sum()

0

In [27]:
final_df.isnull().sum()

ISBN                        0
Book-Title                  0
Book-Author                 1
Year-Of-Publication         0
Publisher                   2
Image-URL-S                 0
Image-URL-M                 0
Image-URL-L                 4
User-ID                     0
Book-Rating                 0
Location                    0
Age                    277835
dtype: int64

In [28]:
final_df.dtypes

ISBN                    object
Book-Title              object
Book-Author             object
Year-Of-Publication     object
Publisher               object
Image-URL-S             object
Image-URL-M             object
Image-URL-L             object
User-ID                  int64
Book-Rating              int64
Location                object
Age                    float64
dtype: object

*NOTE:* We have not removed any of the observations from the dataset because as for now, we don't know which columns we are going to use in the process of model building.

## *Recommendation System Types:*
### 1. Popularity based recomendation:
This system recommends a product/content based on the popularity of that product/content on a particular platform. eg. Top 25 IMDB rated movies, Youtube trending page, etc. It does not consider preferences of user.

### 2. Content/Product based recommendation:
This system recommends content/product based on the previous likes of the user. It considers the preferences of the user.

### 3. Collaborative filtering based recommendation:
This system recommends content/product based on the user ratings. In simple words, consider plotting a graph of ratings for a particular product/content, we find distance (Euclidean, etc) between other product/content in the same plane, the least distance one is recommended.
For more info visit: https://developers.google.com/machine-learning/recommendation/collaborative/basics

### 4. Hybrid based recommendation:
This system considers above all systems and make one recommendation out of them with giving certain weightages.

## Let's make a `Popularity-Based Recommendation`

In [29]:
# We just need `books_df` and `ratings_df` for this.
# Let's start with merging them.
ratings_name_df = ratings_df.merge(books_df, on="ISBN")
ratings_name_df.head()

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
2     6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
3     8680  034545104X            5  Flesh Tones: A Novel  M. J. Rose   
4    10314  034545104X            9  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   
2                2002  Ballantine Books   
3                2002  Ballantine Books   
4                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   
2  http://images.amazon.com/images/P/034545104X.0...   
3  http://images.amazon.com/images/P/034545104X.0...   
4  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...  
2  http://images.amazon.com/images/P/034545104X.0...  
3  http://images.amazon.com/images/P/034545104X.0...  
4  http://images.amazon.com/images/P/034545104X.0...

In [30]:
# We only need `Book-Title` and `Book-Rating` column
titles_ratings_df = ratings_name_df.groupby("Book-Title").count()["Book-Rating"].reset_index()
titles_ratings_df.head()

Book-Title  Book-Rating
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1
3   Ask Lily (Young Women of Faith: Lily Series, ...            1
4   Beyond IBM: Leadership Marketing and Finance ...            1

In [31]:
# Making Num-of-ratings column
titles_ratings_df.rename(columns={"Book-Rating" : "Num-of-ratings"}, inplace=True)
titles_ratings_df

Book-Title  Num-of-ratings
0        A Light in the Storm: The Civil War Diary of ...               4
1                                   Always Have Popsicles               1
2                    Apple Magic (The Collector's series)               1
3        Ask Lily (Young Women of Faith: Lily Series, ...               1
4        Beyond IBM: Leadership Marketing and Finance ...               1
...                                                   ...             ...
241066                                      Ã?Â?lpiraten.               2
241067                     Ã?Â?rger mit Produkt X. Roman.               4
241068                                Ã?Â?sterlich leben.               1
241069                              Ã?Â?stlich der Berge.               3
241070                                  Ã?Â?thique en toc               2

[241071 rows x 2 columns]

In [32]:
# Make avg-ratings dataframe
avg_ratings_df = ratings_name_df.groupby("Book-Title").mean()["Book-Rating"].reset_index()
avg_ratings_df.rename(columns={"Book-Rating" : "Avg-ratings"}, inplace=True)
avg_ratings_df

Book-Title  Avg-ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [33]:
# Merge these dataframes together
popularity_df = titles_ratings_df.merge(avg_ratings_df, on="Book-Title")
popularity_df

Book-Title  Num-of-ratings  \
0        A Light in the Storm: The Civil War Diary of ...               4   
1                                   Always Have Popsicles               1   
2                    Apple Magic (The Collector's series)               1   
3        Ask Lily (Young Women of Faith: Lily Series, ...               1   
4        Beyond IBM: Leadership Marketing and Finance ...               1   
...                                                   ...             ...   
241066                                      Ã?Â?lpiraten.               2   
241067                     Ã?Â?rger mit Produkt X. Roman.               4   
241068                                Ã?Â?sterlich leben.               1   
241069                              Ã?Â?stlich der Berge.               3   
241070                                  Ã?Â?thique en toc               2   

        Avg-ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [34]:
popularity_df["Avg-ratings"].value_counts()

0.000000     105504
5.000000      12922
8.000000      11968
10.000000     11760
7.000000       9357
              ...  
1.593750          1
4.273810          1
3.720930          1
4.016393          1
4.137255          1
Name: Avg-ratings, Length: 3996, dtype: int64

In [35]:
# Let's sort the dataframe using top rated 50 movies by atleast 250+ users
popularity_df = popularity_df[popularity_df["Num-of-ratings"]>=250].sort_values("Avg-ratings", ascending=False).head(50)
popularity_df

Book-Title  Num-of-ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)             428   
80422        Harry Potter and the Goblet of Fire (Book 4)             387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)             278   
80426   Harry Potter and the Order of the Phoenix (Boo...             347   
80414    Harry Potter and the Chamber of Secrets (Book 2)             556   
191612  The Hobbit : The Enchanting Prelude to The Lor...             281   
187377  The Fellowship of the Ring (The Lord of the Ri...             368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...             575   
211384     The Two Towers (The Lord of the Rings, Part 2)             260   
219741                              To Kill a Mockingbird             510   
183573                                  The Da Vinci Code             898   
187880                 The Five People You Meet in Heaven             430   
180556                             The Catcher in the Rye             449   
196326                          The Lovely Bones: A Novel            1295   
764                                                  1984             284   
144165                           Prodigal Summer: A Novel             253   
128670                                         Neverwhere             265   
206502                            The Secret Life of Bees             774   
168719  Stupid White Men ...and Other Sorry Excuses fo...             283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...             493   
204387                The Red Tent (Bestselling Backlist)             723   
191589               The Hitchhiker's Guide to the Galaxy             268   
129379   Nickel and Dimed: On (Not) Getting By in America             335   
93381                                       Into the Wild             252   
63867                                      Fahrenheit 451             409   
74750                           Girl with a Pearl Earring             526   
136145                                          Outlander             283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...             585   
156102                     Seabiscuit: An American Legend             275   
107962                                         Life of Pi             664   
176845                                     The Bean Trees             389   
2281    A Child Called \It\": One Child's Courage to S...             265   
8434                                       ANGELA'S ASHES             279   
76343                                         Good in Bed             490   
64931   Fast Food Nation: The Dark Side of the All-Ame...             321   
12700                                       American Gods             302   
161645                                 Skipping Christmas             322   
105777  Left Behind: A Novel of the Earth's Last Days ...             318   
189551    The Golden Compass (His Dark Materials, Book 1)             336   
181679                                   The Color Purple             314   
160336                               Silence of the Lambs             256   
8752                                          About a Boy             262   
158138                  Seven Up (A Stephanie Plum Novel)             278   
175097  The Alchemist: A Fable About Following Your Dream             266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...             269   
170101                       Suzanne's Diary for Nicholas             457   
111073                                  Lord of the Flies             259   
5664                              A Prayer for Owen Meany             413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)             301   
233851       White Oleander : A Novel (Oprah's Book Club)             356   

        Avg-ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
80414    

In [36]:
popularity_df = popularity_df.merge(books_df, on="Book-Title").drop_duplicates("Book-Title")[["Book-Title", "Book-Author", "Num-of-ratings", "Avg-ratings", "Image-URL-M"]]
popularity_df

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                                                1984   
72                            Prodigal Summer: A Novel   
73                                          Neverwhere   
78                             The Secret Life of Bees   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147  Left Behind: A Novel of the Earth's Last Days ...   
149    The Golden Compass (His Dark Materials, Book 1)   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195       White Oleander : A Novel (Oprah's Book Club)   

                Book-Author  Num-of-ratings  Avg-ratings  \
0             J. K. Rowling             428     5.852804   
3             J. K. Rowling             387     5.824289   
5             J. K. Rowling             278     5.737410   
9             J. K. Rowling             347     5.501441   
13            J. K. Rowling             556     5.183453   
16           J.R.R. TOLKIEN             281     5.007117   
17           J.R.R. TOLKIEN             368     4.948370   
26            J. K. Rowling             575     4.895652   
28           J.R.R. TOLKIEN             260     4.880769   
39               Harper Lee             510     4.700000   
47                Dan Brown             898     4.642539   
53              Mitch Albom             430     4.551163   
55            J.D. Salinger             449     4.545657   
62             Alice Sebold            1295     4.468726   
63            George Orwell             284     4.454225   
72       Barbara Kingsolver             253     4.450593   
73              Neil Gaiman             265     4.449057   
78   

In [37]:
popularity_df["Image-URL-M"][0]

'http://images.amazon.com/images/P/0439136350.01.MZZZZZZZ.jpg'

In [38]:
popularity_df["Book-Author"].unique()[:10]

array(['J. K. Rowling', 'J.R.R. TOLKIEN', 'Harper Lee', 'Dan Brown',
       'Mitch Albom', 'J.D. Salinger', 'Alice Sebold', 'George Orwell',
       'Barbara Kingsolver', 'Neil Gaiman'], dtype=object)

## Let's make a `Collaborating-filtering based Recommendation`

In [39]:
ratings_name_df

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

In [40]:
ratings_name_df.groupby("User-ID").count()["Book-Rating"]

User-ID
2          1
8         17
9          3
10         1
12         1
          ..
278846     1
278849     4
278851    23
278852     1
278854     8
Name: Book-Rating, Length: 92106, dtype: int64

In [41]:
# Let's consider users with more than 200 ratings on a book
ratings_name_df.groupby("User-ID").count()["Book-Rating"] > 200

User-ID
2         False
8         False
9         False
10        False
12        False
          ...  
278846    False
278849    False
278851    False
278852    False
278854    False
Name: Book-Rating, Length: 92106, dtype: bool

In [42]:
x = ratings_name_df.groupby("User-ID").count()["Book-Rating"] > 200
users_200plus_ratings = x[x].index

In [43]:
filtered_ratings = ratings_name_df[ratings_name_df["User-ID"].isin(users_200plus_ratings)]
filtered_ratings.shape

(474007, 10)

In [44]:
filtered_ratings.head(10)

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
2      6543  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
5     23768  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
7     28523  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
15    77940  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
16    81977  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
18    98391  034545104X            9  Flesh Tones: A Novel  M. J. Rose   
21   123981  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
27   135045  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
31   145451  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
42   208406  034545104X            6  Flesh Tones: A Novel  M. J. Rose   

   Year-Of-Publication         Publisher  \
2                 2002  Ballantine Books   
5                 2002  Ballantine Books   
7                 2002  Ballantine Books   
15                2002  Ballantine Books   
16                2002  Ballantine Books   
18                2002  Ballantine Books   
21                2002  Ballantine Books   
27                2002  Ballantine Books   
31                2002  Ballantine Books   
42                2002  Ballantine Books   

                                          Image-URL-S  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   
18  http://images.amazon.com/images/P/034545104X.0...   
21  http://images.amazon.com/images/P/034545104X.0...   
27  http://images.amazon.com/images/P/034545104X.0...   
31  http://images.amazon.com/images/P/034545104X.0...   
42  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-M  \
2   http://images.amazon.com/images/P/034545104X.0...   
5   http://images.amazon.com/images/P/034545104X.0...   
7   http://images.amazon.com/images/P/034545104X.0...   
15  http://images.amazon.com/images/P/034545104X.0...   
16  http://images.amazon.com/images/P/034545104X.0...   
18  http://images.amazon.com/images/P/034545104X.0...   
21  http://images.amazon.com/images/P/034545104X.0...   
27  http://images.amazon.com/images/P/034545104X.0...   
31  http://images.amazon.com/images/P/034545104X.0...   
42  http://images.amazon.com/images/P/034545104X.0...   

                                          Image-URL-L  
2   http://images.amazon.com/images/P/034545104X.0...  
5   http://images.amazon.com/images/P/034545104X.0...  
7   http://images.amazon.com/images/P/034545104X.0...  
15  http://images.amazon.com/images/P/034545104X.0...  
16  http://images.amazon.com/images/P/034545104X.0...  
18  http://images.amazon.com/images/P/034545104X.0...  
21  http://images.amazon.com/images/P/034545104X.0...  
27  http://images.amazon.com/images/P/034545104X.0...  
31  http://images.amazon.com/images/P/034545104X.0...  
42  http://images.amazon.com/images/P/034545104X.0...

In [45]:
y = filtered_ratings.groupby("Book-Title").count()["Book-Rating"] >= 50
famous_books = y[y].index

In [46]:
final_ratings = filtered_ratings[filtered_ratings["Book-Title"].isin(famous_books)]
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [47]:
final_ratings.drop_duplicates()

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [50]:
pivot = final_ratings.pivot_table(index="Book-Title", columns="User-ID", values="Book-Rating")
pivot.fillna(0, inplace=True)
pivot

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

In [52]:
# We will find distance between each rows in vector form with the help of cosine-similarity
from sklearn.metrics.pairwise import cosine_similarity
similarity_score = cosine_similarity(pivot)
cosine_similarity(pivot).shape    # This gives Euclidean distance of 706 books with 706 books.

(706, 706)

*NOTE:* To know about `cosine-similarity` click:https://en.wikipedia.org/wiki/Cosine_similarity

In [53]:
similarity_score

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [72]:
# Let's make a recommendation-function
import numpy as np

def recommend(book_name, num_of_recommendations=5):
  """This function takes book_name as an input and gives suggesions as an output based of our saved pivot table.
  """
  index = np.where(pivot.index==book_name)[0][0]    # Getting index of the book name from pivot dataframe.
  similar_books = sorted(list(enumerate(similarity_score[index])), key=lambda x: x[1], reverse=True)[1:num_of_recommendations+1]
  
  recommendations = []

  for i in similar_books:
    item = []
    temp = books_df[books_df["Book-Title"] == pivot.index[i[0]]]
    item.extend(list(temp.drop_duplicates('Book-Title')['Book-Title'].values))
    item.extend(list(temp.drop_duplicates('Book-Title')['Book-Author'].values))
    item.extend(list(temp.drop_duplicates('Book-Title')['Image-URL-M'].values))
    recommendations.append(item)

  return recommendations

In [73]:
recommend("Message in a Bottle", num_of_recommendations=1)

[['Nights in Rodanthe',
  'Nicholas Sparks',
  'http://images.amazon.com/images/P/0446531332.01.MZZZZZZZ.jpg']]